src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs

```diff
@@ -205,9 +206,55 @@ private ImmutableArray<ISymbol> GetSymbolsForTypeOrNamespaceContext()
                 return symbols.WhereAsArray(s => !s.IsDelegateType() && !s.IsInterfaceType());
             }
 
+            if (_context.IsAsyncMemberDeclarationContext)
+            {
+                return symbols.WhereAsArray(IsSymbolValidForAsyncDeclarationContext);
+            }
+
             return symbols;
         }
 
+        private bool IsSymbolValidForAsyncDeclarationContext(ISymbol symbol)
+        {
+            if (symbol.IsNamespace())
+            {
+                return true;
+            }
+
+            if (symbol is not INamedTypeSymbol namedType ||
+                symbol.IsDelegateType() ||
+                namedType.IsEnumType())
+            {
+                return false;
+            }
+
+            if (namedType.TypeKind == TypeKind.Interface)
+            {
+                // The only interfaces, that are valid in async context are IAsyncEnumerable and IAsyncEnumerator.
+                // So if we are validating an interface, then we can just check for 2 of this possible variants
+                return (namedType.Name == "IAsyncEnumerable" || namedType.Name == "IAsyncEnumerator") &&
+                    namedType.TypeParameters.Length == 1 &&
+                    namedType.IsFromSystemRuntimeOrMscorlibAssembly();
+            }
+
+            if (namedType.TypeKind == TypeKind.Class &&
+                namedType.Name == "Task" &&
+                namedType.TypeParameters.Length < 2 &&
+                namedType.IsFromSystemRuntimeOrMscorlibAssembly())
+            {
+                return true;
+            }
+
+            var attributes = namedType.GetAttributes();
+
+            if (attributes.Any(el => el.AttributeClass?.Name == "AsyncMethodBuilderAttribute" && el.AttributeClass!.IsFromSystemRuntimeOrMscorlibAssembly()))
```

Comment:
1. nameof?
2. @genlu  how risky is this in terms of perf.  We'll have to crack all attributes on all symbols in order to do this.  that could be quite expensive.  do we have precedence for this anywhere?

Reply:
1. Done
2. This is the core logic of this PR. If we refuse to do this, there will be no sense in the whole PR. I didn't experience any performance issues in debigging mode, so, it should be even better, when built in release with all optimizations